cmd/moby/docker.go

```diff
@@ -20,91 +20,89 @@ import (
+		os.Stdin = origStdin
+	}()
 
-	stderr, err := ioutil.ReadAll(stderrPipe)
+	// write input to a file
+	rawInput, err := ioutil.ReadAll(input)
 	if err != nil {
-		return []byte{}, err
+		return []byte{}, errors.New("could not read input")
 	}
-
-	err = cmd.Wait()
-	if err != nil {
-		return []byte{}, fmt.Errorf("%v: %s", err, stderr)
+	tmpInputFileName := "tmpInput"
+	if err := ioutil.WriteFile(tmpInputFileName, rawInput, 0644); err != nil {
+		return []byte{}, errors.New("could not stage input to file")
 	}
-
-	log.Debugf("docker run: %s...Done", strings.Join(args, " "))
-	return stdout, nil
-}
-
-func dockerRunInput(input io.Reader, args ...string) ([]byte, error) {
-	log.Debugf("docker run (input): %s", strings.Join(args, " "))
-	docker, err := exec.LookPath("docker")
+	inputFile, err := os.Open(tmpInputFileName)
 	if err != nil {
-		return []byte{}, errors.New("Docker does not seem to be installed")
+		return []byte{}, errors.New("could not open input file for container stdin")
 	}
-	args = append([]string{"run", "--rm", "-i"}, args...)
-	cmd := exec.Command(docker, args...)
-	cmd.Stdin = input
+	defer os.Remove(tmpInputFileName)
 
-	stderrPipe, err := cmd.StderrPipe()
-	if err != nil {
-		return []byte{}, err
-	}
+	os.Stdin = inputFile
 
-	stdoutPipe, err := cmd.StdoutPipe()
+	var resp container.ContainerCreateCreatedBody
+	resp, err = cli.ContainerCreate(context.Background(), &container.Config{
+		Image:        image,
+		Cmd:          args,
+		AttachStdout: true,
+		AttachStdin:  true,
+	}, &container.HostConfig{
+		AutoRemove: true,
+		LogConfig:  container.LogConfig{Type: "none"},
+	}, nil, "")
 	if err != nil {
+		// if the image wasn't found, try to do a pull and another create
+		if client.IsErrNotFound(err) {
+			if err := dockerPull(image, false); err != nil {
+				return []byte{}, err
+			}
+			resp, err = cli.ContainerCreate(context.Background(), &container.Config{
+				Image:        image,
+				Cmd:          args,
+				AttachStdout: true,
+				AttachStdin:  true,
+			}, &container.HostConfig{
+				AutoRemove: true,
+				LogConfig:  container.LogConfig{Type: "none"},
+			}, nil, "")
+			// if we error again, bail
+			if err != nil {
+				return []byte{}, err
+			}
+		}
 		return []byte{}, err
 	}
 
-	err = cmd.Start()
+	hijackedResp, err := cli.ContainerAttach(context.Background(), resp.ID, types.ContainerAttachOptions{
+		Stdout: true,
+		Stdin:  true,
+		Stream: true,
+	})
 	if err != nil {
 		return []byte{}, err
 	}
 
-	stdout, err := ioutil.ReadAll(stdoutPipe)
-	if err != nil {
+	if err := cli.ContainerStart(context.Background(), resp.ID, types.ContainerStartOptions{}); err != nil {
 		return []byte{}, err
 	}
 
-	stderr, err := ioutil.ReadAll(stderrPipe)
-	if err != nil {
+	if _, err = cli.ContainerWait(context.Background(), resp.ID); err != nil {
 		return []byte{}, err
 	}
 
-	err = cmd.Wait()
-	if err != nil {
-		return []byte{}, fmt.Errorf("%v: %s", err, stderr)
-	}
-
+	out, _ := ioutil.ReadAll(hijackedResp.Reader)
```

Comment:
Probably can just stream this out to the caller instead of pulling it all into memory.